src/backend/distributed/ddl/foreign_constraint.c

```diff
@@ -21,11 +22,588 @@
+
+		frelGraph->nodeMap = hash_create("foreign key relation map (oid)",
+										 64 * 32, &info, hashFlags);
+	}
+
+	fkeyRel = heap_open(ConstraintRelationId, AccessShareLock);
+	fkeyScan = systable_beginscan(fkeyRel, ConstraintRelidIndexId, true,
+								  NULL, 0, NULL);
+
+	while (HeapTupleIsValid(tuple = systable_getnext(fkeyScan)))
+	{
+		Form_pg_constraint con = (Form_pg_constraint) GETSTRUCT(tuple);
+		bool referringFound = false;
+		bool referredFound = false;
+		FRelNode *referringNode = NULL;
+		FRelNode *referredNode = NULL;
+		Oid referringRelationOid = InvalidOid;
+		Oid referredRelationOid = InvalidOid;
+
+		/* Not a foreign key */
+		if (con->contype != CONSTRAINT_FOREIGN)
+		{
+			continue;
+		}
+
+		/*
+		 * Check whether we have seen  referring or referred relations in
+		 * constraint table before. If we haven't, add them to the hash map
+		 * of graph.
+		 */
+		referringRelationOid = con->conrelid;
+		referredRelationOid = con->confrelid;
+
+		referringNode = (FRelNode *) hash_search(frelGraph->nodeMap,
+												 &referringRelationOid,
+												 HASH_ENTER, &referringFound);
+		referredNode = (FRelNode *) hash_search(frelGraph->nodeMap, &referredRelationOid,
+												HASH_ENTER, &referredFound);
+
+		/*
+		 * If we found a node in the graph we only need to add referred node to
+		 * the adjacency  list of that node.
+		 */
+		if (referringFound)
+		{
+			/*
+			 * If referred node is already in the adjacency list of referred node, do nothing.
+			 */
+			if (referringNode->adjacencyList == NIL || !list_member(
+					referringNode->adjacencyList, referredNode))
+			{
+				/*
+				 * If referred node also exists, add it to the adjacency list
+				 * and continue.
+				 */
+				if (referredFound)
+				{
+					referringNode->adjacencyList = lappend(referringNode->adjacencyList,
+														   referredNode);
+				}
+				else
+				{
+					/*
+					 * We need to setup the node and add it to the adjacency list
+					 * of referring node.
+					 */
+					referredNode->adjacencyList = NIL;
+					referredNode->index = curIndex++;
+					referringNode->adjacencyList = lappend(referringNode->adjacencyList,
+														   referredNode);
+				}
+			}
+			else
+			{
+				continue;
+			}
+		}
+		else
+		{
+			/*
+			 * If referring node is not exist in the graph, set its remaining parameters.
+			 */
+			referringNode->adjacencyList = NIL;
+			referringNode->index = curIndex++;
+			if (referredFound)
+			{
+				referringNode->adjacencyList = lappend(referringNode->adjacencyList,
+													   referredNode);
+			}
+			else
+			{
+				/*
+				 * We need to setup the node and add it to the adjacency list
+				 * of referring node.
+				 */
+				referredNode->adjacencyList = NIL;
+				referredNode->index = curIndex++;
+				referringNode->adjacencyList = lappend(referringNode->adjacencyList,
+													   referredNode);
+			}
+		}
+	}
+
+
+	frelGraph->nodeCount = curIndex;
```

Comment:
Strictly, this is unnecessary; the hash knows how many nodes there are: `hash_get_num_entries`